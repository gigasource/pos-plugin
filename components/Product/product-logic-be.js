import { watch } from 'vue'
import cms from 'cms'
import {
  products,
  categories
} from './product-logic'
import {
  currentAppType
} from '../AppType'
import { ObjectID } from 'bson'
import _ from 'lodash'

watch(() => currentAppType.value, async () => {
  await loadCategories()
  await loadProducts()
})

const Product = cms.getModel('Product')
const Category = cms.getModel('Category')

export async function loadProducts() {
  products.value = await Product.find({ appType: currentAppType.value })
}

export async function loadCategories() {
  categories.value = await Category.find({ appType: currentAppType.value })
}

const colors = ['#FFFFFF','#CE93D8','#B2EBF2','#C8E6C9','#DCE775','#FFF59D','#FFCC80','#FFAB91']

export async function createProduct(product) {
  product.category = product.category.map(category => {
    if (category._id) return category._id
    return category
  })
  product.appType = currentAppType.value
  if (!product._id) {
    product._id = new ObjectID()
  }
  if (!product.id) {
    const maxId = products.value.reduce((maxId, product) => {
      return Math.max(maxId, parseInt(product.id))
    }, 0)
    product.id = (maxId ? maxId + 1 : 0)
  }
  // fixme: Add product layout color here if specified
  // { layouts[0].color = Math.floor(Math.random() * colors.length) }
  products.value.push(product)
  await Product.create(product)
  return product
}

export async function updateProductInfo(product) {
  product.category = product.category.map(category => {
    if (category._id) return category._id
    return category
  })
  const foundProduct = products.value.find(_product => _product._id.toString() === product._id.toString())
  Object.assign(foundProduct, { ...product })
  await Product.updateOne({ _id: product._id }, product)
}

export async function removeProductInfo(productId) {
  _.remove(products.value, _product => _product._id.toString() === productId.toString())
  await Product.remove({ _id: productId })
}

export async function createCategory(newCategory) {
  if (!newCategory._id)
    newCategory._id = new ObjectID()
  if (!!categories.value.find(category => category.name === newCategory.name)) return
  newCategory.appType = currentAppType.value
  categories.value.push(newCategory)
  await Category.create(newCategory)
}

export async function updateCategory(_id, change) {
  const foundCategory = categories.value.find(_category => _category._id.toString() === _id.toString())
  if (!foundCategory) return
  Object.assign(foundCategory, change)
  await Category.findOneAndUpdate({_id}, change)
}

export async function updateAllCategory(newCategories) {
  for (const category of newCategories) {
    if (!category._id) category._id = new ObjectID()
    category.appType = currentAppType.value
    const foundCategory = categories.value.find(_category => _category._id.toString() === category._id.toString())
    if (foundCategory)
      Object.assign(foundCategory, category)
    else
      categories.value.push(category)
    await Category.updateOne({ _id: category._id }, category, { upsert: true })
  }
}

export async function deleteCategory(categoryId) {
  _.remove(categories.value, category => category._id.toString() === categoryId.toString())
  await Category.remove({ _id: categoryId })
}
