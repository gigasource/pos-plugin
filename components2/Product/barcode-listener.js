import _ from 'lodash'
export const barcodeHook = new (require('schemahandler/hooks/hooks'))();

let inputString = ''

const debounceFn = _.debounce(() => {
	if (inputString.length >= 10) {
		barcodeHook.emit('newBarcode', inputString)
	}
	inputString = ''
}, 90)

function handleKeyup(e) {
	// check keyboard event is number or alphabet character
	if ((e.which >= 48 && e.which <= 57) || (e.which >= 65 && e.which <= 90) || (e.which >= 97 && e.which <= 122)) {
		inputString += String.fromCharCode(e.which)
	} else {
		return e.preventDefault()
	}
	debounceFn()
}

export function setHandleKeyupBarcode() {
	document.addEventListener('keyup', handleKeyup)
}

export function removeHandleKeyupBarcode() {
	document.removeEventListener('keyup', handleKeyup)
}
