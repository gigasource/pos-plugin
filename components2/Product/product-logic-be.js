import { watch } from 'vue'
import cms from 'cms'
import {
  products,
  categories
} from './product-logic'
import {
  currentAppType
} from '../AppType'
import { ObjectID } from 'bson'
import _ from 'lodash'

watch(() => currentAppType.value, async () => {
  await loadCategories()
  await loadProducts()
})

const Product = cms.getModel('Product')
const Category = cms.getModel('Category')

export async function loadProducts() {
  products.value = await Product.find({ appType: currentAppType.value })
}

export async function loadCategories() {
  categories.value = await Category.find({ appType: currentAppType.value })
}

export async function createProduct(product) {
  product.category = product.category.map(category => {
    if (category._id) return category._id
    return category
  })
  product.appType = currentAppType.value
  if (!product._id) {
    product._id = new ObjectID()
  }
  if (!product.id) {
    const maxId = products.value.reduce((maxId, product) => {
      return Math.max(maxId, parseInt(product.id))
    }, 0)
    product.id = (maxId ? maxId + 1 : 0)
  }
  products.value.push(product)
  await Product.create(product)
  return product
}

export async function updateProductInfo(product) {
  product.category = product.category.map(category => {
    if (category._id) return category._id
    return category
  })
  const foundProduct = products.value.find(_product => _product._id.toString() === product._id.toString())
  Object.assign(foundProduct, { ...product })
  await Product.updateOne({ _id: product._id }, product)
}

export async function removeProductInfo(productId) {
  _.remove(products.value, _product => _product._id.toString() === productId.toString())
  await Product.remove({ _id: productId })
}

export async function createCategory(newCategory) {
  if (!newCategory._id)
    newCategory._id = new ObjectID()
  if (!!categories.value.find(category => category.name === newCategory.name)) return
  categories.value.push(newCategory)
  await Category.create(newCategory)
}

export async function updateCategory(_id, change) {
  await Category.findOneAndUpdate({_id}, change)
}

export async function deleteCategory(categoryId) {
  await Category.remove({ _id: categoryId })
}
