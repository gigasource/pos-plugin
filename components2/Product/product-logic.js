import { ref, watchEffect, computed } from 'vue'

export const products = ref([])
export const categories = ref([])

/**
 * If category.available is true, then this category can be deleted
 */
watchEffect(() => {
  categories.value.forEach(category => {
    const product = products.value.find(product => {
      return !!product.category.find(categoryId => {
        return categoryId === category._id.toString()
      })
    })
    category.available = !product
    // depth max is 2
    // If category is not available (can not be deleted) so is the parent
    if (!category.available && category.parentCategory) {
      const parentCategory = categories.value.find(_category => {
        return _category._id.toString() === category.parentCategory.toString()
      })
      parentCategory.available = false
    }
  })
})

/**
 * Format category to format with subCategory as array attribute
 * {
 *   ...
 *   subCategory: list of sub categories
 *   ...
 * }
 */
export const formattedCategories = computed(() => {
  const result = categories.value.map(category => {
    category.subCategory = _.cloneDeep(categories.value.filter(_category => {
      return _category.parentCategory && _category.parentCategory.toString() === category._id.toString()
    }))
    category.subCategory = !!category.subCategory ? category.subCategory : []
    return category
  })
  _.remove(result, category => {
    return !!category.parentCategory
  })
  return result
})

