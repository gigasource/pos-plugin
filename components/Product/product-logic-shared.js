import cms from 'cms'
import {loadProducts} from "./product-logic-be";

cms.socket.on('commit:handler:finish:Product', async () => {
	await loadProducts()
})
