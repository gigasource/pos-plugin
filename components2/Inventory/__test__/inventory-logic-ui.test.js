import {
  inventories,
  filter,
  filteredInventory,
  addFilter,
  clearFilter,
  removeFilter
} from "../inventory-logic-ui";
import { products, categories } from "../../Product/product-logic";
import { nextTick } from "vue";

const moment = require("moment");

/**
 * All date will be 05/01/2021
 */
function genDate() {
  return moment("05.01.2021", "DD.MM.YYYY").toDate();
}

/**
 * In this test, inventoryHistories lacks some attributes
 */
describe("Test inventory logic ui", function() {
  beforeAll(async () => {
    inventories.value.push({
      _id: "1",
      productId: "1",
      unit: "piece",
      stock: 30
    });
    inventories.value.push({
      _id: "2",
      productId: "2",
      unit: "kg",
      stock: 12
    });
    inventories.value.push({
      _id: "3",
      productId: "3",
      unit: "piece",
      stock: 50
    });
    inventories.value.push({
      _id: "4",
      productId: "4",
      unit: "piece",
      stock: 15
    });
    products.value.push({
      _id: "1",
      category: ["1"],
      name: "Fish"
    });
    products.value.push({
      _id: "2",
      category: ["1"],
      name: "Meat"
    });
    products.value.push({
      _id: "3",
      category: ["2"],
      name: "Apple"
    });
    products.value.push({
      _id: "4",
      category: ["2"],
      name: "Banana"
    });
    categories.value.push({
      _id: "1",
      name: "Food"
    });
    categories.value.push({
      _id: "2",
      name: "Fruit"
    });
    categories.value.push({
      _id: "3",
      name: "Cream"
    });
  });

  it("Case 1: Test filter inventory", async () => {
    addFilter({
      category: {
        name: "Food"
      }
    });
    await nextTick();
    expect(filter.value).toMatchInlineSnapshot(`
      Object {
        "category": Object {
          "name": "Food",
        },
      }
    `);
    expect(filteredInventory.value).toMatchInlineSnapshot(`
      Array [
        Object {
          "_id": "1",
          "product": Object {
            "_id": "1",
            "category": Array [
              Object {
                "_id": "1",
                "available": false,
                "name": "Food",
              },
            ],
            "name": "Fish",
          },
          "productId": "1",
          "stock": 30,
          "unit": "piece",
        },
        Object {
          "_id": "2",
          "product": Object {
            "_id": "2",
            "category": Array [
              Object {
                "_id": "1",
                "available": false,
                "name": "Food",
              },
            ],
            "name": "Meat",
          },
          "productId": "2",
          "stock": 12,
          "unit": "kg",
        },
      ]
    `);
    addFilter({
      name: "Fish"
    });
    await nextTick();
    expect(filter.value).toMatchInlineSnapshot(`
      Object {
        "category": Object {
          "name": "Food",
        },
        "name": "Fish",
      }
    `);
    expect(filteredInventory.value).toMatchInlineSnapshot(`
      Array [
        Object {
          "_id": "1",
          "product": Object {
            "_id": "1",
            "category": Array [
              Object {
                "_id": "1",
                "available": false,
                "name": "Food",
              },
            ],
            "name": "Fish",
          },
          "productId": "1",
          "stock": 30,
          "unit": "piece",
        },
      ]
    `);
    clearFilter();
    await nextTick();
    expect(filter.value).toMatchInlineSnapshot(`Object {}`);
    addFilter({
      stock: [12, 31]
    });
    await nextTick();
    expect(filter.value).toMatchInlineSnapshot(`
      Object {
        "stock": Array [
          12,
          31,
        ],
      }
    `);
    expect(filteredInventory.value).toMatchInlineSnapshot(`
      Array [
        Object {
          "_id": "1",
          "product": Object {
            "_id": "1",
            "category": Array [
              Object {
                "_id": "1",
                "available": false,
                "name": "Food",
              },
            ],
            "name": "Fish",
          },
          "productId": "1",
          "stock": 30,
          "unit": "piece",
        },
        Object {
          "_id": "2",
          "product": Object {
            "_id": "2",
            "category": Array [
              Object {
                "_id": "1",
                "available": false,
                "name": "Food",
              },
            ],
            "name": "Meat",
          },
          "productId": "2",
          "stock": 12,
          "unit": "kg",
        },
        Object {
          "_id": "4",
          "product": Object {
            "_id": "4",
            "category": Array [
              Object {
                "_id": "2",
                "available": false,
                "name": "Fruit",
              },
            ],
            "name": "Banana",
          },
          "productId": "4",
          "stock": 15,
          "unit": "piece",
        },
      ]
    `);
    removeFilter("stock");
    await nextTick();
    expect(filter.value).toMatchInlineSnapshot(`Object {}`);
    expect(filteredInventory.value).toMatchInlineSnapshot(`
      Array [
        Object {
          "_id": "1",
          "product": Object {
            "_id": "1",
            "category": Array [
              Object {
                "_id": "1",
                "available": false,
                "name": "Food",
              },
            ],
            "name": "Fish",
          },
          "productId": "1",
          "stock": 30,
          "unit": "piece",
        },
        Object {
          "_id": "2",
          "product": Object {
            "_id": "2",
            "category": Array [
              Object {
                "_id": "1",
                "available": false,
                "name": "Food",
              },
            ],
            "name": "Meat",
          },
          "productId": "2",
          "stock": 12,
          "unit": "kg",
        },
        Object {
          "_id": "3",
          "product": Object {
            "_id": "3",
            "category": Array [
              Object {
                "_id": "2",
                "available": false,
                "name": "Fruit",
              },
            ],
            "name": "Apple",
          },
          "productId": "3",
          "stock": 50,
          "unit": "piece",
        },
        Object {
          "_id": "4",
          "product": Object {
            "_id": "4",
            "category": Array [
              Object {
                "_id": "2",
                "available": false,
                "name": "Fruit",
              },
            ],
            "name": "Banana",
          },
          "productId": "4",
          "stock": 15,
          "unit": "piece",
        },
      ]
    `);
  });
});
