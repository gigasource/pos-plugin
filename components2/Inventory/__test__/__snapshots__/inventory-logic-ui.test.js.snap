// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test inventory logic ui Case 2: Test inventory histories 1`] = `
Array [
  Object {
    "amount": 30,
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "3",
    "type": "add",
  },
  Object {
    "amount": 10,
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "3",
    "type": "add",
  },
  Object {
    "amount": 10,
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "3",
    "type": "remove",
  },
  Object {
    "amount": 25,
    "date": 2021-01-04T17:00:00.000Z,
    "inventory": "2",
    "type": "add",
  },
]
`;

exports[`Test inventory logic ui Case 2: Test inventory histories 2`] = `
Array [
  Object {
    "_id": "2",
    "add": 25,
    "category": Object {
      "name": "Food",
    },
    "history": Array [
      Object {
        "amount": 25,
        "date": 2021-01-04T17:00:00.000Z,
        "inventory": "2",
        "type": "add",
      },
    ],
    "inventory": "2",
    "name": "Meat",
    "remove": 0,
    "stock": 12,
    "unit": "kg",
  },
  Object {
    "_id": "3",
    "add": 40,
    "category": Object {
      "name": "Fruit",
    },
    "history": Array [
      Object {
        "amount": 30,
        "date": 2021-01-04T17:00:00.000Z,
        "inventory": "3",
        "type": "add",
      },
      Object {
        "amount": 10,
        "date": 2021-01-04T17:00:00.000Z,
        "inventory": "3",
        "type": "add",
      },
      Object {
        "amount": 10,
        "date": 2021-01-04T17:00:00.000Z,
        "inventory": "3",
        "type": "remove",
      },
    ],
    "inventory": "3",
    "name": "Apple",
    "remove": 10,
    "stock": 50,
    "unit": "piece",
  },
]
`;
